### PR TITLE
api: using index to replace tree scan if only returns count 

### LINKED ISSUE #5607: Add filter in /pd/api/v1/stats/region API
### Describe your feature request related problem

TiDB need to get region count between `startKey` and `endKey` from `/pd/api/v1/stats/region` to calculate TiFlash replica progress. `GetRegionStatus` will scan regions in `BTree` so it will have heavy cost if there are many regions.
In fact, [TiDB only need region count](https://github.com/pingcap/tidb/blob/dc223253f0cb1e5101ddf92b677e190da292811d/ddl/ddl_tiflash_api.go#L387), if `/pd/api/v1/stats/region` can add parameter to only get region count, the cost of `/pd/api/v1/stats/region` will reduce from O(N) to O(logN).

1 million regions, call `/pd/api/v1/stats/region` per 2 seconds, flame graph on PD:
7bae2d26-3aa0-43a9-b86f-dc7f79e39a7d

### Describe the feature you'd like
Add a parameter in `/pd/api/v1/stats/region` to get only region count.

### PR DESCRIPTION
**Pull request by bufferflies**
### What problem does this PR solve?

Issue Number: Close #5607

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test

Code changes

- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))

Side effects

Related changes

### Release note

```release-note
 None.
```

### CODE REVIEW

File: server/core/region.go [hunk 1]
```diff
@@ -1160,6 +1161,27 @@ func (r *RegionInfo) GetWriteLoads() []float64 {
...
+	var endIndex int
+	var item rangetree.RangeItem
+	// it should return the length of the tree if endKey is nil.
+	if len(endKey) == 0 {
+		endIndex = r.tree.tree.Len() - 1
+	} else {
+		item, endIndex = r.tree.tree.GetWithIndex(end)
+		// it should return the endIndex - 1 if the endKey is the startKey of a region.
+		if item != nil && bytes.Equal(item.GetStartKey(), endKey) {
+			endIndex--
```
Comment (JmPotato) [inline server/core/region.go:1179]: Why not just return 1 here to make it more clear?

**Comment by bufferflies** [comment]
the fixed pr flame:
image

File: server/core/region.go (hunk 1 above)
Comment (bufferflies) [reply server/core/region.go:1179]: It needs to consider the start key index if the endKey is equal the item startKey.

Comment (JmPotato) [reply server/core/region.go:1193]: Do we have a test case to test this kind of situation?

**Review by JmPotato — COMMENTED** [review]
The rest LGTM.

File: server/core/region.go (hunk 1 above)
Comment (bufferflies) [reply server/core/region.go:1193]: yes， the first case(startkey="",endKey="") can be found in the stats_test.go.

**Review by bufferflies — COMMENTED** [review]
Yes, the ("","") will be found in the api unit test.

File: server/core/region.go [hunk 2]
```diff
@@ -1174,6 +1175,27 @@ func (r *RegionInfo) GetWriteLoads() []float64 {
...
+func (r *RegionsInfo) GetRangeCount(startKey, endKey []byte) int {
+	start := &regionItem{&RegionInfo{meta: &metapb.Region{StartKey: startKey}}}
+	end := &regionItem{&RegionInfo{meta: &metapb.Region{StartKey: endKey}}}
+	// it returns 0 if startKey is nil.
+	_, startIndex := r.tree.tree.GetWithIndex(start)
+	var endIndex int
+	var item rangetree.RangeItem
+	// it should return the length of the tree if endKey is nil.
+	if len(endKey) == 0 {
+		endIndex = r.tree.tree.Len() - 1
```
Comment (CabinfeverB) [inline server/core/region.go:1188]: If start equals "", it seems return wrong ans?

Comment (bufferflies) [reply server/core/region.go:1188]: No, the first region key range is start with "".

**Review by CabinfeverB — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**